### PR TITLE
[FSDP] Added private `_unshard` API

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124304

Some toy example:
Screenshot 2024-04-17 at 2 00 05 PM

We define `FullyShardedDataParallel._unshard(async_op: bool = False)` that can be used to prefetch all-gathers. The user should make sure:
1. Run lazy init before the first `_unshard` call of training. For example, this can hackily be done via `root_module.check_is_root()` on the root FSDP module `root_module`.
2. Call `root_module._wait_unshard_streams_on_current_stream()` before the first `_unshard` call of the current iteration (just need to call it once after last optimizer step and before first `_unshard` of this iteration).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Differential Revision: [D56262876](https://our.internmc.facebook.com/intern/diff/D56262876)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56262876).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56262876).

File: test/distributed/fsdp/test_fsdp_comm.py [hunk 1]
```diff
@@ -283,7 +288,102 @@ def reset_mocks():
...
+    def test_unshard_async(self, use_orig_params: bool):
+        class ReduceModule(nn.Module):
+            def __init__(self, dim: int, group: dist.ProcessGroup):
+                super().__init__()
+                self.group = group
+                self.weight = nn.Parameter(torch.randn(dim, dim))
+
+            def forward(self, x: torch.Tensor):
+                y = F.relu(x @ self.weight)
+                # NOTE: This all-reduce is not differentiable and is included
```
Comment (wanchaol) [inline test/distributed/fsdp/test_fsdp_comm.py:307]: oh iiuc this allreduce would simply be ignored by autograd and that's why we can still do loss.backward() later?

File: test/distributed/fsdp/test_fsdp_comm.py [hunk 2]
```diff
@@ -283,7 +288,102 @@ def reset_mocks():
...
+
+        for _ in range(10):
+            losses: List[torch.Tensor] = []
+            for _model, _optim in ((ref_model, ref_optim), (model, optim)):
+                losses.append(_model(inp).sum())
+                losses[-1].backward()
+                _optim.step()
+                _optim.zero_grad()
+            self.assertEqual(losses[0], losses[1])
+            model.module.mlps._wait_unshard_streams_on_current_stream()
```
Comment (wanchaol) [inline test/distributed/fsdp/test_fsdp_comm.py:382]: how could we tell in the test that the allreduce is overlapped with the unshard? or did you check this in trace?

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for adding this!

File: test/distributed/fsdp/test_fsdp_comm.py (hunk 2 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_comm.py:382]: had to check in the trace manually since it is hard to check it programmatically
